src/main.cpp

```diff
@@ -2968,6 +3043,95 @@ bool static LoadBlockIndexDB()
+        return false;
+    }
+    return true;
+}
+
+bool UndoFileIsOpenable(int nFile)
+{
+    CDiskBlockPos pos(nFile, 0);
+    if (!vinfoBlockFile[nFile].nUndoSize)
+        return false;
+    if (CAutoFile(OpenUndoFile(pos, true), SER_DISK, CLIENT_VERSION).IsNull()) {
+        ClearUndoFileInfo(nFile);
+        return false;
+    }
+    return true;
+}
+
+bool DataFilesAreOpenable(int nFile)
+{
+    if (BlockFileIsOpenable(nFile) && UndoFileIsOpenable(nFile))
+        return true;
+    return false;
+}
+
+int LastBlockInFile(int nFile)
+{
+    return vinfoBlockFile[nFile].nHeightLast;
+}
+
+
+bool CheckBlockFiles()
+{
+    LOCK(cs_main);
+    // Check presence of blk files
+    int nKeepMinBlksFromHeight = nPrune ? (max((int)(chainActive.Height() - MIN_BLOCKS_TO_KEEP), 0)) : 0;
+    LogPrintf("Checking all required data for active chain is available (mandatory from height %i to %i)\n", nKeepMinBlksFromHeight, max(chainActive.Height(), 0));
+    set<int> setDataPruned, setUndoPruned;
+    setDataFilePrunable.clear();
+    setUndoFilePrunable.clear();
+    for (CBlockIndex* pindex = chainActive.Tip(); pindex && pindex->pprev; pindex = pindex->pprev) {
+        if (pindex->nHeight > nKeepMinBlksFromHeight) {
+            if (!(pindex->nStatus & BLOCK_HAVE_DATA) || !(pindex->nStatus & BLOCK_HAVE_UNDO)) { // Fail immediately if required data is missing
+                LogPrintf("Error: Required data for block: %i is missing.\n", pindex->nHeight);
+                return false;
+            } else if (!DataFilesAreOpenable(pindex->nFile)) { // Or if data is unreadable
+                LogPrintf("Error: Required file for block: %i can't be opened.\n", pindex->nHeight);
+                return false;
+            }
+        } else { // Check consistency and pruneability of unrequired data
+            if (pindex->nStatus & BLOCK_HAVE_DATA) {
+                if (BlockFileIsOpenable(pindex->nFile)) {
+                    if (chainActive.Height() > AUTOPRUNE_AFTER_HEIGHT && LastBlockInFile(pindex->nFile) < nKeepMinBlksFromHeight) { // Mark pruneable data
+                        setDataFilePrunable.insert(pindex->nFile);
+                    }
+                } else {
+                    pindex->nStatus &= ~BLOCK_HAVE_DATA;
+                    setDirtyBlockIndex.insert(pindex);
+                }
+            }
+            if (pindex->nStatus & BLOCK_HAVE_UNDO) {
+                if (UndoFileIsOpenable(pindex->nFile)) {
+                    if (chainActive.Height() > AUTOPRUNE_AFTER_HEIGHT && LastBlockInFile(pindex->nFile) < nKeepMinBlksFromHeight) { // Mark pruneable data
+                        setUndoFilePrunable.insert(pindex->nFile);
+                    }
+                } else {
+                    pindex->nStatus &= ~BLOCK_HAVE_UNDO;
+                    setDirtyBlockIndex.insert(pindex);
+                }
+            }
+        }
+        if (~pindex->nStatus & BLOCK_HAVE_DATA && pindex->nStatus & BLOCK_VALID_CHAIN)
```

Comment:
This shouldn't require BLOCK_VALID_CHAIN anymore; I think BLOCK_VALID_TRANSACTIONS is enough.